tools/clang/lib/CodeGen/CGStmt.cpp

```diff
@@ -1202,11 +1202,19 @@ void CodeGenFunction::EmitBreakStmt(const BreakStmt &S) {
 
   // HLSL Change Begin - incorporate unconditional branch blocks into loops
   // If it has a continue location, it's a loop
-  if (BreakContinueStack.back().ContinueBlock.getBlock() && (BreakContinueStack.size() < 2 ||
-      BreakContinueStack.back().ContinueBlock.getBlock() != BreakContinueStack.end()[-2].ContinueBlock.getBlock())) {
-    assert(EHStack.getInnermostActiveNormalCleanup() == EHStack.stable_end() && "HLSL Shouldn't need cleanups");
-    CGM.getHLSLRuntime().EmitHLSLCondBreak(*this, CurFn, BreakContinueStack.back().BreakBlock.getBlock(),
-                                           BreakContinueStack.back().ContinueBlock.getBlock());
+  llvm::BasicBlock *lastContinueBlock = BreakContinueStack.back().ContinueBlock.getBlock();
+  if (lastContinueBlock && (BreakContinueStack.size() < 2 ||
+      lastContinueBlock != BreakContinueStack.end()[-2].ContinueBlock.getBlock())) {
+    // We execute this if
+    // - we are in an unnested loop, or
+    // - we are in a nested loop but the continue block of the enclosing loop is different from the current continue block.
+    // TODO: Is the second conditional even possible in a language without goto or "continue N"?
```

Comment:
I suppose I'd like @pow2clk to comment here.

Reply:
Correct the second case to be "we are in a nested control *control construct* but the continue block of the enclosing loop is different from the current continue block"

In order for the TODO to make sense to me, I think it's asking how that it is possible that a nested loop have the same continue as the outer loop. It seems trivially easy for a nested loop to have a different continue as the outer loop.

The answer to the quandry I altered so that I have an interesting answer to is: a switch statement (control construct) inside a loop. In this case, since the switch needs a break location, but doesn't have a continue location that is unique from the enclosing loop, clang copies the continue from the containing loop since any continue found within the switch should point to the continue for that outer loop. Previously this confused the implementation and added dx.breaks for switch statements, which wasn't necessary.

I admit I failed to document this properly. There is mention of it in the description of this commit: https://github.com/microsoft/DirectXShaderCompiler/pull/2795/commits/6ea8dc00ec0d662db23a10cfc43d6abf28885eca which is the only way I was able to remember what was going on. I should have added a comment to the code. I was young and foolish when I wrote this.

Reply:
Fixed the comment.
Thanks for the explanation @pow2clk! I missed the fact that this doesn't have to be nested loops but other constructs can have break or continue locations, too.
Removed the TODO, this was purely for my own understanding :). Do you want me to remove the rest of the comment, too, or replace it with more explanation while we're at it?